type generated by template disappears in C code.

When I generate a type in a template, everything works fine, but when I use the exact same template to generate another type and use that type, this type will disappear in the generated C code, and the generated C code will use the wrong type instead. Here is an example.

```nim
template foobar(arg: typed): untyped =
  type
    MyType = object
      member: type(arg)

  var myVar: MyType
  myVar.member = arg
  echo myVar


foobar("hello world")
foobar(123.456'f64)
```

This is the relevant generated C code:
```C

// there is only one instance of MyType
struct tyObject_MyType_c8zLdw0BfP5QV35I5Dp9cLQ {
NimStringDesc* member;
};
// an instance of MyType with a member of type double does not exist.

/*...*/

// as you can see both myVar instances have the same type
tyObject_MyType_c8zLdw0BfP5QV35I5Dp9cLQ myVar_JAYCzYxe9bQSb28Ss2x9aoxg;
tyObject_MyType_c8zLdw0BfP5QV35I5Dp9cLQ myVar_JAYCzYxe9bQSb28Ss2x9aoxg_2;

/*...*/

N_LIB_PRIVATE N_NIMCALL(void, NimMainModule)(void) {
	NimStringDesc* T1_;
	tyArray_nHXaesL0DJZHyVS07ARPRA T2_;
	tyArray_nHXaesL0DJZHyVS07ARPRA T3_;
nimRegisterGlobalMarker(TM_X9bGGOoNyKehOTF7f2xys3g_2);
	T1_ = (NimStringDesc*)0;
	T1_ = myVar_JAYCzYxe9bQSb28Ss2x9aoxg.member; myVar_JAYCzYxe9bQSb28Ss2x9aoxg.member = copyStringRC1(((NimStringDesc*) &TM_X9bGGOoNyKehOTF7f2xys3g_3));
	if (T1_) nimGCunrefNoCycle(T1_);
	memset((void*)T2_, 0, sizeof(tyArray_nHXaesL0DJZHyVS07ARPRA));
	T2_[0] = dollar__cL9bBANF2jNLmbCKCDXwjDQ(myVar_JAYCzYxe9bQSb28Ss2x9aoxg);
	echoBinSafe(T2_, 1);
        // here is the line that won't compile
	myVar_JAYCzYxe9bQSb28Ss2x9aoxg_2.member = 1.2345600000000000e+02;
	memset((void*)T3_, 0, sizeof(tyArray_nHXaesL0DJZHyVS07ARPRA));
	T3_[0] = dollar__cL9bBANF2jNLmbCKCDXwjDQ_2(myVar_JAYCzYxe9bQSb28Ss2x9aoxg_2);
	echoBinSafe(T3_, 1);
}
```
The C compiler will refuse to compile with this error message:

```
/tmp/nimcache/scratch.c:269:42: error: incompatible types when assigning to type 'NimStringDesc * {aka struct NimStringDesc *}' from type 'double'
  myVar_JAYCzYxe9bQSb28Ss2x9aoxg_2.member = 1.2345600000000000e+02;
```

Because of weak type checking in C compiling integers will silently work, so ``foobar(123)`` will compile just fine. But when the generated instance of MyType has a member of an integer type, casting pointers to it will actually work, but it will cause a segfault at runtime, when the integer is too small to hold an actual pointer value.

Interesting though, there are two different `$` procedures generated for the same type. You can see them in use in the generated code as ``dollar__cL9bBANF2jNLmbCKCDXwjDQ`` and ``dollar__cL9bBANF2jNLmbCKCDXwjDQ_2``

--- a/compiler/sighashes.nim
+++ b/compiler/sighashes.nim
@@ -193,16 +193,15 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]) =
         # generated object names can be identical, so we need to
         # disambiguate furthermore by hashing the field types and names:
         # mild hack to prevent endless recursions (makes nimforum compile again):
-        let wasAnon = sfAnon in t.sym.flags
-        excl t.sym.flags, sfAnon
+        let oldFlags = t.sym.flags
+        t.sym.flags = t.sym.flags - {sfAnon, sfGenSym}
         let n = t.n
         for i in 0 ..< n.len:
           assert n[i].kind == nkSym
           let s = n[i].sym
           c.hashSym s
           c.hashType s.typ, flags
-        if wasAnon:
-          incl t.sym.flags, sfAnon
+        t.sym.flags = oldFlags
     else:
       c &= t.id
     if t.len > 0 and t.sons[0] != nil:

--- a/compiler/sighashes.nim
+++ b/compiler/sighashes.nim
@@ -189,10 +189,11 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]) =
         c.hashTypeSym(t.sym)
       else:
         c.hashSym(t.sym)
-      if sfAnon in t.sym.flags:
+      if {sfAnon, sfGenSym} * t.sym.flags != {}:
         # generated object names can be identical, so we need to
         # disambiguate furthermore by hashing the field types and names:
         # mild hack to prevent endless recursions (makes nimforum compile again):
+        let wasAnon = sfAnon in t.sym.flags
         excl t.sym.flags, sfAnon
         let n = t.n
         for i in 0 ..< n.len:
@@ -200,7 +201,8 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]) =
           let s = n[i].sym
           c.hashSym s
           c.hashType s.typ, flags
-        incl t.sym.flags, sfAnon
+        if wasAnon:
+          incl t.sym.flags, sfAnon
     else:
       c &= t.id
     if t.len > 0 and t.sons[0] != nil:

--- a/tests/types/t7905.nim
+++ b/tests/types/t7905.nim
@@ -1,0 +1,18 @@
+discard """
+  output: '''
+(member: "hello world")
+(member: 123.456)
+'''
+"""
+
+template foobar(arg: typed): untyped =
+  type
+    MyType = object
+      member: type(arg)
+
+  var myVar: MyType
+  myVar.member = arg
+  echo myVar
+
+foobar("hello world")
+foobar(123.456'f64)

--- a/tests/types/t7905.nim
+++ b/tests/types/t7905.nim
@@ -2,6 +2,8 @@ discard """
   output: '''
 (member: "hello world")
 (member: 123.456)
+(member: "hello world", x: ...)
+(member: 123.456, x: ...)
 '''
 """
 
@@ -16,3 +18,16 @@ template foobar(arg: typed): untyped =
 
 foobar("hello world")
 foobar(123.456'f64)
+
+template foobarRec(arg: typed): untyped =
+  type
+    MyType = object
+      member: type(arg)
+      x: ref MyType
+
+  var myVar: MyType
+  myVar.member = arg
+  echo myVar
+
+foobarRec("hello world")
+foobarRec(123.456'f64)

